rclcpp/src/rclcpp/generic/generic_subscription.cpp

```diff
@@ -0,0 +1,93 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/generic/generic_subscription.hpp"
+
+#include <memory>
+#include <string>
+
+#include "rcl/subscription.h"
+#include "rclcpp/generic/typesupport_helpers.hpp"
+
+namespace rclcpp
+{
+namespace generic
+{
+
+namespace
+{
+rcl_subscription_options_t get_subscription_options(const rclcpp::QoS & qos)
+{
+  auto options = rcl_subscription_get_default_options();
+  options.qos = qos.get_rmw_qos_profile();
+  return options;
+}
+}  // unnamed namespace
+
+GenericSubscription::GenericSubscription(
+  rclcpp::node_interfaces::NodeBaseInterface * node_base,
+  const std::shared_ptr<rcpputils::SharedLibrary> ts_lib,
+  const std::string & topic_name,
+  const std::string & topic_type,
+  const rclcpp::QoS & qos,
+  std::function<void(std::shared_ptr<rclcpp::SerializedMessage>)> callback)
+: SubscriptionBase(
+    node_base,
+    *rclcpp::generic::get_typesupport_handle(
+      topic_type, "rosidl_typesupport_cpp", ts_lib),
+    topic_name,
+    get_subscription_options(qos),
+    true),
+  callback_(callback)
+{}
+
+std::shared_ptr<void> GenericSubscription::create_message()
+{
+  return create_serialized_message();
+}
+
+std::shared_ptr<rclcpp::SerializedMessage> GenericSubscription::create_serialized_message()
+{
+  return std::make_shared<rclcpp::SerializedMessage>(0);
+}
+
+void GenericSubscription::handle_message(
+  std::shared_ptr<void> & message, const rclcpp::MessageInfo & message_info)
+{
+  (void) message_info;
+  auto typed_message = std::static_pointer_cast<rclcpp::SerializedMessage>(message);
+  callback_(typed_message);
+}
+
+void GenericSubscription::handle_loaned_message(
+  void * message, const rclcpp::MessageInfo & message_info)
+{
+  (void) message;
+  (void) message_info;
```

Comment:
while this worked well in the context of `rosbag2`, I don't think this is appropriate within the public API of `rclcpp`. At a minimum I'd say we should log a warning here, most likely though raise an exception of some sort indicating "not implemented".

Reply:
Thanks, done. What would it take to implement that function, i.e. what is `message`?

Reply:
@Karsten1987 bump

Reply:
There is actually a bit more to it than appropriate for this PR. There are some changes needed to the RMW interface as well as the executor. 
We currently have support for either loaning messages or sending/receiving serialized messages. We don't have the cross-functionality of loaning serialized messages exposed through the RMW.

It looks like there's a first PR for publishing loaned messages at the rosbag2 PR: https://github.com/ros2/rosbag2/pull/664, but I don't see anything related to the subscription side yet.